### ISSUE TITLE
Matrix and Vector4?

### ISSUE THREAD

**Issue by illume** [pygame/pygame#3412]
Should we? How should we?

**Comment by novialriptide** [comment]
I feel like if a user wants to use a matrix or VectorN, shouldn't they use numpy?

**Comment by oddbookworm** [comment]
I actually agree that they could be useful. Numpy is certainly efficient for very large arrays, but for smaller arrays, it's got a lot of overhead and pygame vectors handily beat them. I did some testing that is as fair as I think I can get, and here's my results:

vector dot product
```
import pygame
from time import perf_counter
import numpy

iterations = 10_000_000

vec1 = pygame.Vector2(1, 1)
vec2 = pygame.Vector2(-1, 0)

start = perf_counter()
for _ in range(iterations):
    vec1.dot(vec2)
end = perf_counter()

print(f"Pygame vectors took {end - start} seconds")

vec1 = numpy.array([1, 1])
vec2 = numpy.array([-1, 0])

start = perf_counter()
for _ in range(iterations):
    numpy.dot(vec1, vec2)
end = perf_counter()

print(f"Numpy arrays took {end - start} seconds")
```
with results
```
pygame 2.1.2 (SDL 2.0.16, Python 3.10.5)
Hello from the pygame community. https://www.pygame.org/contribute.html
Pygame vectors took 1.2928906559995994 seconds
Numpy arrays took 12.81941641200001 seconds
```


vector addition
```
import pygame
from time import perf_counter
import numpy

iterations = 10_000_000

vec1 = pygame.Vector2(1, 1)
vec2 = pygame.Vector2(-1, 0)

start = perf_counter()
for _ in range(iterations):
    vec1 + vec2
end = perf_counter()

print(f"Pygame vectors took {end - start} seconds")

vec1 = numpy.array([1, 1])
vec2 = numpy.array([-1, 0])

start = perf_counter()
for _ in range(iterations):
    numpy.add(vec1, vec2)
end = perf_counter()

print(f"Numpy arrays took {end - start} seconds")
```
with results
```pygame 2.1.2 (SDL 2.0.16, Python 3.10.5)
Hello from the pygame community. https://www.pygame.org/contribute.html
Pygame vectors took 1.1397429660000853 seconds
Numpy arrays took 6.464601710000352 seconds
```

**Comment by Starbuck5** [comment]
Matrices and 4D Vectors seem to come up in conversation occasionally. I don't understand the use case for 2D games, but I'm not a very math-y person. Because of that I just see this as extra complexity for pygame, I'm not interested in taking it on. But I don't want to be negative, that's just my opinion, maybe others see a use for this and are interested in working on it.

**Comment by illume** [comment]
> I actually agree that they could be useful. Numpy is certainly efficient for very large arrays, but for smaller arrays, it's got a lot of overhead and pygame vectors handily beat them. I did some testing that is as fair as I think I can get, and here's my results:

Indeed! numpy isn't interested in improving this since the beginning. The library 'numeric' which numpy evolved from was better in this regard. Perhaps it will change of course.

With pygame math.Vector3 /2; I believe we still have a lot of speed left on the table. Because we're not using any sort of object pool (freelist) still AFAIK.

>  I don't understand the use case for 2D games

In 2D games, being able to scale + rotate + translate all at once (cameras) is very useful. This makes the transform matrix is very-very useful for 2d games.

Anything with physics, and of course 2.5d/3d math. Audio and vision processing... the list goes on.

Matrix (3x3 and 4x4) and Vector4 are very useful and commonly used in games and some other classes of apps. Currently people are using other libraries(there are dozens of python libraries existing) for them when needed, avoid using them, or roll their own.

We already have Surfarray which is a specialized a 3/4 x X x Y vector, and Color which is a specialized Vector4.

Teaching Vectors and Matrices is also something people do with python.

**Comment by ThanhChinhBK** [comment]
Could I work on this issue?

**Comment by novialriptide** [comment]
> Could I work on this issue?

Hi, thanks for your interest in contributing to pygame. I think we need a solid plan before we add Matrices and Vector4s to pygame.

If you're finding things to work on, we are currently working on `pygame.geometry` and we would love to have new contributors help out on it! https://github.com/novialriptide/newpygamestuff

**Comment by illume** [comment]
I think Vector4 is much less work, and less controversial. If it's just a Vector4 like Vector2/3 it's probably not super hard. The majority of the work would probably be in the tests. We already have plenty of tests that could be refactored to work with N length vectors rather than individual tests for each vector length. But even the test refactoring work could be done separately to the Vector4 implementation work in C.
